ember-client/src/test/scala/org/http4s/ember/client/internal/ClientHelpersSuite.scala

```diff
@@ -103,127 +101,92 @@ class ClientHelpersSuite extends Http4sSuite {
       .assertEquals(Some(name))
   }
 
-  test("Postprocess response should reuse when body is run") {
```

Comment:
I deleted several tests that relied on the body being run. They can exist in some other form but they aren't applicable here anymore since the body doesn't necessarily need to be run for reuse anymore (i.e. if we already have a chunk with the body).

Reply:
In other words, reuse is dependent on whether the `drain` function returns a `Some`. If it does, it means we've already buffered enough to read the next request/response. If it returns a `None`, that means we don't have enough bytes so close the connection.

Note that for variable-length bodies, if the body hasn't been read, it's going to always close, just on account of how the parser is coded.

Reply:
Will `drain` return `None` if we've successfully consumed the entire body and are simply awaiting the next message?    I'm afraid I may have pointed in the wrong direction when I stated:

> If there are just a few more bytes to read, but they're not here yet, maybe our socket handle is better used elsewhere. I suppose request timeouts deal with this.

It's one thing to aggressively close when we know that we have to discard bytes that haven't arrived yet before we get to the next interesting one.  What I think we don't want is to aggressively close the connection when we successfully drained the previous message.  The non-pipeline case is common, where one response is wholly received before the next request begins.  If we're closing immediately in that case, I think we'll rarely have reused connections and performance will suffer.

I was trying to study prior art, and I think the relevant part of Jetty's implementation is [here](https://github.com/eclipse/jetty.project/blob/29d48b0de5fe4529c6aedb13fe96934147bcf9a8/jetty-server/src/main/java/org/eclipse/jetty/server/HttpConnection.java#L413), but I a lot more context to absorb before fully understanding it.

Reply:
If we've read and consumed all available bytes and are awaiting the next message, `drain` is going to return `Some(Array.emptyByteArray)`. I'll add some tests to exercise that behavior